File: parsers/swift/Sources/Synx/SynxStringify.swift
// Value → SYNX text. Mirrors crates/synx-core/src/lib.rs `serialize`.
import Foundation

public enum SynxStringify {

    public static let maxDepth = 128

    public static func stringify(_ value: SynxValue) -> String {
        var out = ""
        out.reserveCapacity(2048)
        serialize(value, depth: 0, into: &out)
        return out
    }

    private static func serialize(_ v: SynxValue, depth: Int, into out: inout String) {
        if depth > maxDepth { out.append("[synx:max-depth]\n"); return }
        guard case .object(let map) = v else {
            out.append(formatPrimitive(v))
            return
        }
        let indent = String(repeating: " ", count: depth * 2)
        for key in map.keys.sorted() {
            guard let val = map[key] else { continue }
            switch val {
            case .array(let arr):
                out.append(indent); out.append(key); out.append("\n")
                for item in arr {
                    if case .object(let inner) = item {
                        let keys = inner.keys
                        if let first = keys.first, let firstVal = inner[first] {
                            out.append(indent); out.append("  - "); out.append(first)
                            out.append(" "); out.append(formatPrimitive(firstVal)); out.append("\n")
                            for k in keys.dropFirst() {
                                if let v = inner[k] {
                                    out.append(indent); out.append("    "); out.append(k)
                                    out.append(" "); out.append(formatPrimitive(v)); out.append("\n")
                                }
                            }
                        }
                    } else {
                        out.append(indent); out.append("  - ")
                        out.append(formatPrimitive(item)); out.append("\n")
                    }
                }
            case .object:
                out.append(indent); out.append(key); out.append("\n")
                serialize(val, depth: depth + 1, into: &out)
            case .string(let s) where s.contains("\n"):
                out.append(indent); out.append(key); out.append(" |\n")
                for line in s.split(separator: "\n", omittingEmptySubsequences: false) {
                    out.append(indent); out.append("  "); out.append(String(line)); out.append("\n")
                }
            default:
                out.append(indent); out.append(key); out.append(" ")
                out.append(formatPrimitive(val)); out.append("\n")
            }
        }
    }

    public static func formatPrimitive(_ v: SynxValue) -> String {
        switch v {
        case .string(let s): return s
        case .int(let n): return String(n)
        case .float(let f):
            if f.isNaN || f.isInfinite { return "null" }
            var s = String(format: "%.17g", f)
            if !s.contains(".") && !s.contains("e") && !s.contains("E") {
                s.append(".0")
            }
            return s
        case .bool(let b): return b ? "true" : "false"
        case .null: return "null"
        case .array(let arr):
            let parts = arr.map { formatPrimitive($0) }
            return "[" + parts.joined(separator: ", ") + "]"
        case .object: return "[Object]"
        case .secret: return "[SECRET]"
        }
    }
}

File: parsers/java/src/main/java/com/aperturesyndicate/synx/SynxStringify.java
package com.aperturesyndicate.synx;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/** Value → SYNX text. Mirrors {@code serialize} in {@code crates/synx-core/src/lib.rs}. */
public final class SynxStringify {

    private SynxStringify() {}
    public static final int MAX_DEPTH = 128;

    public static String stringify(SynxValue v) {
        StringBuilder sb = new StringBuilder(2048);
        serialize(v, 0, sb);
        return sb.toString();
    }

    private static void serialize(SynxValue v, int depth, StringBuilder out) {
        if (depth > MAX_DEPTH) { out.append("[synx:max-depth]\n"); return; }
        if (!(v instanceof SynxValue.Obj obj)) {
            out.append(formatPrimitive(v));
            return;
        }
        SynxObject map = obj.map();
        String indent = " ".repeat(depth * 2);
        List<String> keys = new ArrayList<>(map.keys());
        Collections.sort(keys);
        for (String key : keys) {
            SynxValue val = map.get(key);
            if (val instanceof SynxValue.Arr a) {
                out.append(indent).append(key).append('\n');
                for (SynxValue item : a.values()) {
                    if (item instanceof SynxValue.Obj inner) {
                        var innerEntries = inner.map().entriesList();
                        if (!innerEntries.isEmpty()) {
                            var first = innerEntries.get(0);
                            out.append(indent).append("  - ").append(first.getKey())
                               .append(' ').append(formatPrimitive(first.getValue())).append('\n');
                            for (int j = 1; j < innerEntries.size(); j++) {
                                var e = innerEntries.get(j);
                                out.append(indent).append("    ").append(e.getKey())
                                   .append(' ').append(formatPrimitive(e.getValue())).append('\n');
                            }
                        }
                    } else {
                        out.append(indent).append("  - ").append(formatPrimitive(item)).append('\n');
                    }
                }
            } else if (val instanceof SynxValue.Obj) {
                out.append(indent).append(key).append('\n');
                serialize(val, depth + 1, out);
            } else if (val instanceof SynxValue.Str s && s.value().indexOf('\n') >= 0) {
                out.append(indent).append(key).append(" |\n");
                for (String line : s.value().split("\n", -1)) {
                    out.append(indent).append("  ").append(line).append('\n');
                }
            } else {
                out.append(indent).append(key).append(' ').append(formatPrimitive(val)).append('\n');
            }
        }
    }

    public static String formatPrimitive(SynxValue v) {
        if (v instanceof SynxValue.Str s) return s.value();
        if (v instanceof SynxValue.Int i) return Long.toString(i.value());
        if (v instanceof SynxValue.Float f) {
            double d = f.value();
            if (Double.isNaN(d) || Double.isInfinite(d)) return "null";
            String s = String.format(java.util.Locale.ROOT, "%.17g", d);
            if (s.indexOf('.') < 0 && s.indexOf('e') < 0 && s.indexOf('E') < 0) s += ".0";
            return s;
        }
        if (v instanceof SynxValue.Bool b) return b.value() ? "true" : "false";
        if (v instanceof SynxValue.Null) return "null";
        if (v instanceof SynxValue.Arr a) {
            List<String> parts = new ArrayList<>();
            for (SynxValue x : a.values()) parts.add(formatPrimitive(x));
            return "[" + String.join(", ", parts) + "]";
        }
        if (v instanceof SynxValue.Obj) return "[Object]";
        if (v instanceof SynxValue.Secret) return "[SECRET]";
        return "";
    }
}
